scripts/chat.py

```diff
@@ -0,0 +1,276 @@
+        if system_message:
+            msg = self.B_SYS + system_message + self.E_SYS
+            self.buffer.extend(tokenizer.convert_tokens_to_ids(msg))
+
+    def onecmd(self, line):
+        # Check if the line starts with a special command prefix
+        if line.startswith("/"):
+            command = line[1:].strip().split(" ")[0]
+            args = line[len(command) + 2 :].strip()
+            try:
+                func = getattr(self, "do_" + command)
+                return func(args)
+            except AttributeError:
+                print(f"Unknown command: {line}")
+                return
+        # handles ctrl-D
+        elif line == "EOF":
+            self.do_EOF()
+        else:
+            return self.default(line)
+
+    def default(self, line):
+        if not len(line.strip()):
+            return
+
+        self.buffer.extend(
+            tokenizer.convert_tokens_to_ids(
+                tokenizer.tokenize(self.B_INST + line + self.E_INST)
+            )
+        )
+        ids = torch.tensor(self.buffer, dtype=torch.long, device=device)
+
+        result = generate(
+            model,
+            ids,
+            max_new_tokens=self.max_new_tokens,
+            use_cache=args.no_use_cache,
+            do_sample=not args.deterministic,
+            eos_token_id=tokenizer.eos_token_id,
+        )
+        if local_rank != 0:
+            return
+
+        result = result[len(self.buffer) :].tolist()
+        self.buffer = self.buffer + result + [tokenizer.eos_token_id]
+        print(
+            tokenizer.convert_tokens_to_string(tokenizer.convert_ids_to_tokens(result))
+        )
+
+    def do_set(self, arg):
+        args = re.split("=|\\s", arg)
+        if len(args) != 2:
+            print("usage: /set max_new_tokens=100")
+        if args[0] == "max_new_tokens":
+            self.max_new_tokens = int(args[1])
+        else:
+            print("Invalid attribute: " + args[0])
+
+    def do_buffer(self, arg):
+        print(f"The look-back buffer has {len(self.buffer)} tokens:")
+        print(
+            tokenizer.convert_tokens_to_string(
+                tokenizer.convert_ids_to_tokens(self.buffer)
+            )
+        )
+
+    def do_help(self, arg):
+        print("Help on using the REPL:")
+        print("/help - Show this help message.")
+        print(
+            "/set max_new_tokens <value> - set the maximum length of response. Responses may be truncated."
+        )
+        print("/buffer - show the full buffer of chat history being fed to the model.")
+        print("Any other input is treated as text to be handled by the LLM.")
+        print("ctrl-c or ctrl-d to exit.")
+
+    def do_quit(self, line=""):
+        self.do_EOF(line)
+
+    def do_exit(self, line=""):
+        self.do_EOF(line)
+
+    def do_EOF(self, line=""):
+        print("Exiting...")
+        exit()
+
+
+if __name__ == "__main__":
+    try:
+        ChatREPL(args.max_new_tokens).cmdloop()
```

Comment:
maybe add a cmd arg or a file option to load a system prompt?

Reply:
there's a cli arg for this at the top (it's `arg....`) and you can also override it while running with `/set max_new_tokens=xxx`

Reply:
I see `max_new_tokens`, but not `system_prompt`

Reply:
ah I misunderstood, sure I can add that too